core/src/bootstrap.rs

```diff
@@ -0,0 +1,766 @@
+    // We're going to mint 6 distinct types of tokens and create the pool and refresh boxes as
+    // described in EIP-23. The minting of each type of token requires a distinct transaction, so we
+    // need 8 transactions in total. We assume that the resulting token-holding boxes generated from
+    // these transactions each has a box value of `erg_value_per_box`. Similarly the pool and
+    // refresh boxes will also hold `erg_value_per_box`.
+    //
+    // Now define `E_i = i*(erg_value_per_box + tx_free)` for `i = 1,2,.., 8`. `E_i` represents the
+    // amount of ERGs necessary to effect `i` remaining transactions.
+    //
+    // So we require a total ERG value of `E_8 = 8*(erg_value_per_box + tx_free)`
+    //
+    // The chain transaction is structured as follows:
+    //   * First sweep the unspent boxes of the wallet for a target balance of `E_8`. Denote these
+    //     input boxes by `I_1`.
+    //
+    //   * Mint the first token with `I_1` as input, resulting in two output boxes:
+    //      - `B_1_token` containing the minted token and `ergo_value_per_box`
+    //      - `B_1_remaining` containing `E_7` in ERG value.
+    //
+    //   * Mint the second token with input boxes containing `B_1_remaining`, resulting in two
+    //     output boxes:
+    //      - `B_2_token` containing the minted token and `ergo_value_per_box`
+    //      - `B_2_remaining` containing `E_6` in ERG value.
+    //
+    // And so on.
+
+    // This variable represents the index `i` described above.
+    let mut num_transactions_left = 8;
+
+    let wallet_pk_ergo_tree = config
+        .addresses
+        .wallet_address_for_chain_transaction
+        .script()?;
+    let guard = wallet_pk_ergo_tree.clone();
+
+    // Since we're building a chain of transactions, we need to filter the output boxes of each
+    // constituent transaction to be only those that are guarded by our wallet's key.
+    let filter_tx_outputs = move |outputs: Vec<ErgoBox>| -> Vec<ErgoBox> {
+        outputs
+            .clone()
+            .into_iter()
+            .filter(|b| b.ergo_tree == guard)
+            .collect()
+    };
+
+    // This closure computes `E_{num_transactions_left}`.
+    let calc_target_balance = |num_transactions_left| {
+        let b = erg_value_per_box.checked_mul_u32(num_transactions_left)?;
+        let fees = tx_fee.checked_mul_u32(num_transactions_left)?;
+        b.checked_add(&fees)
+    };
+
+    // Effect a single transaction that mints a token with given details, as described in comments
+    // at the beginning. By default it uses `wallet_pk_ergo_tree` as the guard for the token box,
+    // but this can be overriden with `different_token_box_guard`.
+    let mint_token = |input_boxes: Vec<ErgoBox>,
+                      num_transactions_left: &mut u32,
+                      token_name,
+                      token_desc,
+                      token_amount,
+                      different_token_box_guard: Option<ErgoTree>|
```

Comment:
Used closures here so that I didn't have to keep passing in common parameters like `erg_value_per_box` and `tx_fee`.